app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,340 @@
+    struct active_tap_dance *tap_dance = find_tap_dance(event.position);
+    if (tap_dance == NULL) {
+        LOG_ERR("ACTIVE TAP DANCE CLEARED TOO EARLY");
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    tap_dance->is_pressed = false;
+    // An active tap dance has already been decided and is registering keypresses.
+    // Release tap dance on binding release.
+    if (tap_dance->tap_dance_decided) {
+        release_tap_dance_behavior(tap_dance, event.timestamp);
+        tap_dance->tap_dance_decided = false;
+    }
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+// behavior_tap_dance_timer_handler
+
+// Determines what happens when a tap-dance's timer expires
+void behavior_tap_dance_timer_handler(struct k_work *item) {
+    // Timer for tap dance has ran out.
+    struct active_tap_dance *tap_dance = CONTAINER_OF(item, struct active_tap_dance, release_timer);
+    if (tap_dance->position == ZMK_BHV_TAP_DANCE_POSITION_FREE) {
+        return;
+    }
+    if (tap_dance->timer_cancelled) {
+        tap_dance->timer_cancelled = false;
+    } else {
+        press_tap_dance_behavior(tap_dance, tap_dance->release_at);
+        tap_dance->tap_dance_decided = true;
+        // Tap dance is currently pressed, prepare for release of binding
+        // before actual behavior release
+        if (tap_dance->is_pressed)
+            return;
+        // Tap dance is not currently pressed. Release the tap dance.
+        LOG_DBG("Tap dance releases immediately! Counter reached: %d", tap_dance->counter);
+        release_tap_dance_behavior(tap_dance, tap_dance->release_at);
+        return;
+    }
+}
+
+static const struct behavior_driver_api behavior_tap_dance_driver_api = {
+    .binding_pressed = on_tap_dance_binding_pressed,
+    .binding_released = on_tap_dance_binding_released,
+};
+
+// tap_dance_position_state_changed_listener
+
+// Declared here, defined below. Determines what happens when another
+// keycode is pressed other than any active tap-dances. When any
+// tap-dance is interrupted, they will resolve before the interruption
+// keycode and wait to release when their binding is released
+static int tap_dance_position_state_changed_listener(const zmk_event_t *eh);
+
+ZMK_LISTENER(behavior_tap_dance, tap_dance_position_state_changed_listener);
+ZMK_SUBSCRIPTION(behavior_tap_dance, zmk_position_state_changed);
+
+static int tap_dance_position_state_changed_listener(const zmk_event_t *eh) {
+    // A keypress with a different position than the last tap dance has been recorded.
+    struct zmk_position_state_changed *ev = as_zmk_position_state_changed(eh);
+    if (ev == NULL) {
+        return ZMK_EV_EVENT_BUBBLE;
+    }
+    for (int i = 0; i < ZMK_BHV_TAP_DANCE_MAX_HELD; i++) {
+        struct active_tap_dance *tap_dance = &active_tap_dances[i];
+        if (tap_dance->position == ZMK_BHV_TAP_DANCE_POSITION_FREE) {
+            continue;
+        }
+        if (tap_dance->position != ev->position && tap_dance->timer_started) {
```

Comment:
if you invert the check in this if to the following, the code has a more consistent style
```
if (tap_dance->position == ev->position && tap_dance->timer_started) {
    continue;
}
```